### PR TITLE
Runtime config

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### Description

We add an initial structure for runtime config. In this PR, we plan to add only the invalid tx configuration for the starter.

### Issue Link

#1636 

### Type of change

New feature (non-breaking change which adds functionality)

### Decision

- Allow Geth to take invalid tx. The config doesn't affect geth util

### TODO

- [x] Fix invalid tx test
- [ ] Add tx validity check. (Will continue #1740)

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit.rs [hunk 1]
```diff
@@ -390,12 +403,17 @@ impl<F: Field> Circuit<F> for EvmCircuit<F> {
...
                     u16_table,
+                    feature_config: params,
                 },
             ),
             challenges,
         )
     }
 
+    fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+        Self::configure_with_params(meta, Default::default())
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit.rs:414]: halo2 will not use this method when `configure_with_params` is implemented.  I would suggest adding an `unreachable!()` here to make sure this method is not called by accident (and thus guarantee that we're always using the features passed by the user).

File: zkevm-circuits/src/evm_circuit.rs [hunk 2]
```diff
@@ -457,7 +478,7 @@ mod evm_circuit_stats {
     fn evm_circuit_unusable_rows() {
         assert_eq!(
             EvmCircuit::<Fr>::unusable_rows(),
-            unusable_rows::<Fr, EvmCircuit::<Fr>>(()),
+            unusable_rows::<Fr, EvmCircuit::<Fr>>(FeatureConfig::default()),
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit.rs:481]: I wonder if different feature configs could lead to different unusable rows :thinking:

Comment (ed255) [inline zkevm-circuits/src/evm_circuit.rs:481]: I wonder if different feature configs could lead to different unusable rows :thinking:

File: zkevm-circuits/src/evm_circuit/param.rs [hunk 3]
```diff
@@ -173,7 +174,14 @@ lazy_static::lazy_static! {
 }
 fn get_step_height_map() -> HashMap<ExecutionState, usize> {
     let mut meta = ConstraintSystem::<Fr>::default();
-    let circuit = EvmCircuit::configure(&mut meta);
+    let circuit = EvmCircuit::configure_with_params(
+        &mut meta,
+        FeatureConfig {
+            // Enable invalid_tx to get ExecutionState height
+            invalid_tx: true,
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/param.rs:181]: What happens when we use this `step_height_map` with the feature `invalid_tx: false`?  Does any state height change?  Or is it only that we have an extra one for the InvalidTx state?
I think we could have the situation where a a state height changes, maybe not now?

I think it's worth adding a check to make sure that all common states between `step_height_map` built out of different `FeatureConfig`s have the same height, to detect bugs as soon as possible.

As an example. with the `invalid_tx` feature, the EndTx constraints change a bit.  Imagine this change causes an extra cell to be allocated, and then we have that `height[EndTx]=5` with `invalid_tx=false`, and `height[EndTx]=6` with `invalid_tx=true`!  Provably this is not the case right now, but I imagine this situation could happen with other features.

File: zkevm-circuits/src/super_circuit.rs [hunk 4]
```diff
@@ -407,19 +401,12 @@ impl<F: Field> Circuit<F> for SuperCircuit<F> {
...
-                max_txs: params.max_txs,
-                max_withdrawals: params.max_withdrawals,
-                max_calldata: params.max_calldata,
-                mock_randomness: params.mock_randomness,
-            },
-        )
+        Self::Config::new(meta, params)
     }
 
     fn configure(_meta: &mut ConstraintSystem<F>) -> Self::Config {
```
Comment (ed255) [inline zkevm-circuits/src/super_circuit.rs:412]: Note: Previously I suggested doing `unreachable!()` in the `configure` of the `EvmCircuit`, it's following this pattern.

File: zkevm-circuits/src/test_util.rs [hunk 5]
```diff
@@ -175,12 +188,19 @@ impl<const NACC: usize, const NTX: usize> CircuitTestBuilder<NACC, NTX> {
...
 
-        let circuit = EvmCircuitCached::get_test_circuit_from_block(block);
-        let prover = MockProver::<Fr>::run(k, &circuit, vec![]).map_err(|err| {
-            CircuitTestError::SynthesisFailure {
-                circuit: Circuit::EVM,
-                reason: err,
-            }
+        // Mainnet EVM circuit constraints can be cached for test performance.
+        // No cache for EVM circuit with customized features
+        let prover = if block.feature_config.is_mainnet() {
```
Comment (ed255) [inline zkevm-circuits/src/test_util.rs:193]: Very nice! This resolves my question about using the cached version in non-mainnet features :)

**Review by ed255 — APPROVED** [review]
Overall looks good to me!  I think the way the runtime config is implemented is very nice; in particular it's not very intrusive!  And the features are now available anywhere where they are needed; to allow the implementation to take different paths as necessary without making thinks too complex :)

I'm approving already even though I left some comments and suggestions; as I think they are minor and don't affect the current implementation.  The most serious one is the state height map discussion.

File: zkevm-circuits/src/evm_circuit.rs [hunk 6]
```diff
@@ -310,6 +314,7 @@ pub(crate) mod cached {
         /// Circuit configuration.  These values are calculated just once.
         static ref CACHE: Cache = {
             let mut meta = ConstraintSystem::<Fr>::default();
+            // Cached EVM circuit is configured with Mainnet FeatureConfig
```
[the remark this answers is not in the document]
Comment (ed255) [reply zkevm-circuits/src/evm_circuit.rs:317]: Self resolved with https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1728#discussion_r1464777767

File: bus-mapping/src/circuit_input_builder.rs [hunk 7]
```diff
@@ -741,7 +790,13 @@ impl<P: JsonRpcClient> BuilderClient<P> {
         prev_state_root: Word,
     ) -> Result<CircuitInputBuilder<FixedCParams>, Error> {
         let block = Block::new(self.chain_id, history_hashes, prev_state_root, eth_block)?;
-        let mut builder = CircuitInputBuilder::new(sdb, code_db, block, self.circuits_params);
+        let mut builder = CircuitInputBuilder::new(
+            sdb,
+            code_db,
+            block,
+            self.circuits_params,
+            FeatureConfig::default(),
```
[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply bus-mapping/src/circuit_input_builder.rs]: I think this is a good idea. Resolved in [f9e197d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1728/commits/f9e197dbdcba9400564f81f78e0c3f2c52b52a87)

File: zkevm-circuits/src/evm_circuit.rs (hunk 1 above)
Comment (ChihChengLiang) [reply zkevm-circuits/src/evm_circuit.rs]: Fixed here. [8787b3c](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1728/commits/8787b3c321647294fc03bf58d61c1b8bcb613dbd)

File: bus-mapping/src/circuit_input_builder.rs [hunk 8]
```diff
@@ -630,6 +680,23 @@ impl<P: JsonRpcClient> BuilderClient<P> {
             cli: client,
             chain_id: chain_id.into(),
             circuits_params,
+            feature_config: FeatureConfig::default(),
```
Comment (Brechtpd) [inline bus-mapping/src/circuit_input_builder.rs:683]: Maybe this should just call through to `new_with_features` with `FeatureConfig::default()`?

File: zkevm-circuits/src/evm_circuit/execution.rs [hunk 9]
```diff
@@ -1259,7 +1276,11 @@ impl<F: Field> ExecutionConfig<F> {
             ExecutionState::BeginTx => assign_exec_step!(self.begin_tx_gadget),
             ExecutionState::EndTx => assign_exec_step!(self.end_tx_gadget),
             ExecutionState::EndBlock => assign_exec_step!(self.end_block_gadget),
-            ExecutionState::InvalidTx => assign_exec_step!(self.invalid_tx),
+            ExecutionState::InvalidTx => {
+                if let Some(invalid_tx) = &self.invalid_tx {
+                    assign_exec_step!(invalid_tx)
+                }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/execution.rs:1282]: Maybe can just unwrap it here and let it error out if it's not set up so it doesn't silently skip it?

**Review by Brechtpd — APPROVED** [review]
Very nice! Only some really unimportant comments from me, the step heights issue pointed out by Edu I think indeed potentially problematic.

File: bus-mapping/src/circuit_input_builder.rs (hunk 8 above)
Comment (ChihChengLiang) [reply bus-mapping/src/circuit_input_builder.rs]: Good idea. Fixed in [949c228](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1728/commits/949c2284162023737ee90d681fe33d71bd220920)

File: zkevm-circuits/src/evm_circuit/execution.rs (hunk 9 above)
Comment (ChihChengLiang) [reply zkevm-circuits/src/evm_circuit/execution.rs]: Nice catch! I fixed it here. [e4be4d0](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1728/commits/e4be4d03736fcd9f80848b4db131181b8b5d1b00)

**Review by ed255 — APPROVED** [review]
I've reviewed the latest changes on this PR and all looks good to me!

**Pull request merged**